server/routes/referrals/referralDataPresenterUtils.test.ts

```diff
@@ -0,0 +1,252 @@
+        it('returns an empty string', () => {
+          const referral = draftReferralFactory.build({ additionalNeedsInformation: null })
+          const userInputData = { 'additional-needs-information': '' }
+          const utils = new ReferralDataPresenterUtils(referral, userInputData)
+
+          expect(utils.stringValue('additionalNeedsInformation', 'additional-needs-information')).toBe('')
+        })
+      })
+
+      describe('when there is user input data and the user input value for the property is a non-empty string', () => {
+        it('returns the user input value', () => {
+          const referral = draftReferralFactory.build({ additionalNeedsInformation: null })
+          const userInputData = { 'additional-needs-information': 'bar' }
+          const utils = new ReferralDataPresenterUtils(referral, userInputData)
+
+          expect(utils.stringValue('additionalNeedsInformation', 'additional-needs-information')).toBe('bar')
+        })
+      })
+
+      describe('when there is user input data and the user input value for the property is present but neither a string nor null', () => {
+        it('returns the stringified user input value', () => {
+          const referral = draftReferralFactory.build({ additionalNeedsInformation: null })
+          const userInputData = { 'additional-needs-information': 100 }
+          const utils = new ReferralDataPresenterUtils(referral, userInputData)
+
+          expect(utils.stringValue('additionalNeedsInformation', 'additional-needs-information')).toBe('100')
+        })
+      })
+    })
+
+    describe('when the referral has a non-null value for the property', () => {
+      describe('when there is no user input data', () => {
+        it('returns the value from the referral', () => {
+          const referral = draftReferralFactory.build({ additionalNeedsInformation: 'foo' })
+          const utils = new ReferralDataPresenterUtils(referral, null)
+
+          expect(utils.stringValue('additionalNeedsInformation', 'additional-needs-information')).toBe('foo')
+        })
+      })
+
+      describe('when there is user input data and the user input value for the property is null', () => {
+        it('returns an empty string', () => {
+          const referral = draftReferralFactory.build({ additionalNeedsInformation: 'foo' })
+          const userInputData = { 'additional-needs-information': null }
+          const utils = new ReferralDataPresenterUtils(referral, userInputData)
+
+          expect(utils.stringValue('additionalNeedsInformation', 'additional-needs-information')).toBe('')
+        })
+      })
+
+      describe('when there is user input data and the property is absent from the user input data', () => {
+        it('returns an empty string', () => {
+          const referral = draftReferralFactory.build({ additionalNeedsInformation: 'foo' })
+          const userInputData = {}
```

Comment:
this is the first case that made me double check. what's the rationale that this empty object overwrites the field?

Reply:
The idea is that if the presenter is displaying the page in response to the user submitting a form (which would usually mean that there's been a validation failure and the user needs to correct some of the data they entered), then instead of presenting the data from the model, it should present the data which the user submitted in the form.

The specific case you're asking about - the user submitted a form but that particular field is absent in the request body - could occur for example with checkboxes, where if the user unticks all of the checkboxes for a field `foo` then the browser will not submit a `foo` key in the request body. We still want to display an empty list of checkboxes in that case. For the example here (a text value) it's a bit of an edge case and would only happen in a request from a browser in the case of e.g. a user managing to uncheck an already-selected radio button. A bit pathological but I thought worth sticking a test in for it.